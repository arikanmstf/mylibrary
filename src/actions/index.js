export function selectBook (book) {
	// selectBook is an Action creator and needs to return an action
	// an object with a type property
	return {
		type : "BOOK_SELECTED",
		payload : book
	};
}