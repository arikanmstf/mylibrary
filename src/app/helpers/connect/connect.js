// @flow
import React, { ComponentType } from 'react';

const connect = (mapUiToProps: Function) => (
  (WrappedComponent: ComponentType<*>) => {
    const uiComponents = mapUiToProps();
    return ({
      location, navigation, match, ...other
    }: Object) => (
      <WrappedComponent
        {...uiComponents}
        {...other}
        navigation={navigation}
        location={location || { pathname: navigation ? navigation.state.routeName : null }}
        match={match || { params: { id: navigation ? navigation.getParam('id') : undefined } }}
      />
    );
  }
);

export default connect;
