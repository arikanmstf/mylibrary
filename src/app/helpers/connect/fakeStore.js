class Store {
  constructor(props) {
    this.props = props;
  }

  toJS() {
    return this.props;
  }
}

class FakeStore {
  constructor(props) {
    this.props = props;
  }

  getState() {
    return new Store(this.props);
  }

  subscribe() {}
}

export const createFakeStore = (data) => {
  return new FakeStore(data);
};

export const dispatch = (action) => (action && typeof action === 'function' ? action() : undefined);
