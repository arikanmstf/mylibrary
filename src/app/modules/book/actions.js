/**
 * Actions Template By => create-module script
 * @version 1.1.0
 *
 */

// @flow
import { createAction } from 'redux-actions';
import { BOOK_UPDATE_BOOK, BOOK_UPDATE_CARD } from 'constants/actions/actionNames';
import { LOADER_CARD_DETAIL } from 'constants/actions/loaderNames';
import logger from 'helpers/logger';
import { showLoader, hideLoader } from 'ui/ModalLoader/actions';

import type { Dispatch } from 'redux';
import type { ThunkAction } from 'redux-thunk';
import type { BookDetail } from 'helpers/api/types';

import { transformBookToCard } from './transformers';
import { getBookDetail } from './services';

export const updateBookAction = createAction(BOOK_UPDATE_BOOK);
export const updateBookCardAction = createAction(BOOK_UPDATE_CARD);

export const updateBook = (book: BookDetail): ThunkAction => {
  return (dispatch: Dispatch<*>) => {
    dispatch(updateBookAction(book));
    dispatch(updateBookCardAction(transformBookToCard(book)));
  };
};

export const fetchBook = (id: number, shouldShowLoader: boolean = true): ThunkAction => {
  return async (dispatch: Dispatch<*>) => {
    if (shouldShowLoader) {
      dispatch(showLoader(LOADER_CARD_DETAIL));
    }

    logger.log('action: fetchBookStart');

    const book = await getBookDetail(dispatch)({ id });
    logger.log('action: fetchBook');

    await Promise.all([
      dispatch(updateBook(book)),
    ]);

    logger.log('action: fetchBookEnd');
    dispatch(hideLoader(LOADER_CARD_DETAIL));
  };
};
