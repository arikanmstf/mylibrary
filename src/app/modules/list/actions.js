/**
 * Actions Template By => create-module script
 * @version 1.1.0
 *
 */

// @flow
import logger from 'helpers/logger';
import { updateRows } from 'modules/row/actions';
import { showLoader, hideLoader } from 'ui/ModalLoader/actions';
import { LOADER_ADD_LIST } from 'constants/actions/loaderNames';

import type { Dispatch } from 'redux';
import type { ThunkAction } from 'redux-thunk';
import type { SubmitSearchListFormRequest } from 'screens/publicationAddToList/PublicationAddToListTypes';
import type { AddListFormRequest } from 'screens/publicationAddToList/addListForm/AddListFormTypes';

import { getLists, postList } from './services';

export const fetchLists = (params: SubmitSearchListFormRequest): ThunkAction => {
  return async (dispatch: Dispatch<*>) => {
    logger.log('action: fetchListsStart');

    const result = await getLists(dispatch)(params);
    logger.log('action: fetchLists', result);

    await dispatch(updateRows(result));
    logger.log('action: fetchListsEnd');
  };
};

export const addList = (params: AddListFormRequest): ThunkAction => {
  return async (dispatch: Dispatch<*>, getState: Function) => {
    logger.log('action: addListStart');
    dispatch(showLoader(LOADER_ADD_LIST));

    const result = await postList(dispatch)(params);

    logger.log('action: addList', result);
    const { rows } = getState().toJS().row;
    const fixedRows = rows || [];
    fixedRows.push(result);

    await dispatch(updateRows(fixedRows));
    dispatch(hideLoader(LOADER_ADD_LIST));
    logger.log('action: addListEnd');
  };
};
