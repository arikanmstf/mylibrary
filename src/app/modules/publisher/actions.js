/**
 * Actions Template By => create-module script
 * @version 1.1.0
 *
 */

// @flow

import { createAction } from 'redux-actions';
import { PUBLISHER_UPDATE_CARD, PUBLISHER_UPDATE_PUBLISHER } from 'constants/actions/actionNames';
import logger from 'helpers/logger';
import { showLoader, hideLoader } from 'ui/ModalLoader/actions';
import { transformPublisherToCard } from 'helpers/data/transform';

import type { Dispatch } from 'redux';
import type { ThunkAction } from 'redux-thunk';

import { getPublisherDetail } from './services';

export const updatePublisher = createAction(PUBLISHER_UPDATE_PUBLISHER);
export const updatePublisherCard = createAction(PUBLISHER_UPDATE_CARD);

export const fetchPublisher = (id: number): ThunkAction => {
  return async (dispatch: Dispatch<*>) => {
    dispatch(showLoader());

    logger.log('action: fetchPublisherStart');

    const publisher = await getPublisherDetail(dispatch)({ id });
    const card = transformPublisherToCard(publisher);
    logger.log('action: fetchPublisher');

    await Promise.all([
      dispatch(updatePublisherCard(card)),
      dispatch(updatePublisher(publisher)),
    ]);

    logger.log('action: fetchPublisherEnd');
    dispatch(hideLoader());
  };
};
