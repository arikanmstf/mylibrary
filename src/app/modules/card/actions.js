// @flow
import { createAction } from 'redux-actions';
import {
  ADD_CARDS,
  UPDATE_CARDS,
  UPDATE_TOTAL_PAGES,
  UPDATE_CURRENT_PAGE,
  UPDATE_SEARCH_QUERY,
  UPDATE_SEARCH_PENDING,
  UPDATE_LIST_TYPE,
} from 'constants/actions/actionNames';
import logger from 'helpers/logger';
import { showLoader, hideLoader } from 'ui/ModalLoader/actions';
import { showCenterLoader, hideCenterLoader } from 'ui/CenterLoader/actions';
import { getPublicationList } from 'modules/publication/services';

import type { Dispatch } from 'redux';
import type { ThunkAction } from 'redux-thunk';

export const addCards = createAction(ADD_CARDS);
export const updateCards = createAction(UPDATE_CARDS);
export const updateTotalPages = createAction(UPDATE_TOTAL_PAGES);
export const updateCurrentPage = createAction(UPDATE_CURRENT_PAGE);
export const updateSearchQuery = createAction(UPDATE_SEARCH_QUERY);
export const updateListType = createAction(UPDATE_LIST_TYPE);
export const updateSearchPending = createAction(UPDATE_SEARCH_PENDING);

export const fetchAndUpdateCards = (shouldShowLoader: boolean = true): ThunkAction => {
  return async (dispatch: Dispatch<*>, getState: Function) => {
    const page = 1;
    if (shouldShowLoader) {
      dispatch(showLoader());
    }
    logger.log('action: fetchAndUpdateCardsStart');
    const search = getState().toJS().card.searchQuery;
    const type = getState().toJS().card.listType;
    const result = await getPublicationList(dispatch)({ page, search, type });
    logger.log('action: fetchAndUpdateCards');
    await Promise.all([
      dispatch(updateCards(result.content)),
      dispatch(updateTotalPages(result.totalPages)),
      dispatch(updateCurrentPage(page)),
    ]);

    logger.log('action: fetchAndUpdateCardsEnd');
    dispatch(hideLoader());
  };
};

export const fetchAndAddCards = (): ThunkAction => {
  return async (dispatch: Dispatch<*>, getState: Function) => {
    const isPending = getState().toJS().card.isSearchPending;
    if (isPending) {
      return false;
    }

    logger.log('action: fetchAndAddCardsStart');
    const page = getState().toJS().card.currentPage + 1;
    const totalPage = getState().toJS().card.totalPages;
    if (page > totalPage) {
      logger.log('page is too big');
      return false;
    }
    const search = getState().toJS().card.searchQuery;
    const type = getState().toJS().card.listType;

    dispatch(showCenterLoader());
    await dispatch(updateSearchPending(true));
    const result = await getPublicationList(dispatch)({ page, search, type });
    logger.log('action: fetchAndAddCards');
    await Promise.all([
      dispatch(addCards(result.content)),
      dispatch(updateTotalPages(result.totalPages)),
    ]);

    dispatch(updateSearchPending(false));
    dispatch(updateCurrentPage(page));
    dispatch(hideCenterLoader());
    logger.log('action: fetchAndAddCardsEnd');
    return true;
  };
};
