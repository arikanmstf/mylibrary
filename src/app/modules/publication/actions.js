// @flow
import { createAction } from 'redux-actions';
import { PUBLICATION_UPDATE_PUBLICATION } from 'constants/actions/actionNames';
import logger from 'helpers/logger';
import { showLoader, hideLoader } from 'ui/Loader/actions';
import { transformPublicationToCard } from 'helpers/data/transform';
import { findIndexById, cloneObjectArray } from 'helpers/data/array';
import { updateCards, updateCard } from 'modules/card/actions';

import type { Dispatch } from 'redux';
import type { ThunkAction } from 'redux-thunk';
import { getPublicationDetail, postToggleList } from './services';
import type { ToggleListRequest } from './types';

export const updatePublication = createAction(PUBLICATION_UPDATE_PUBLICATION);

export const fetchPublication = (id: number, shouldShowLoader: boolean = true): ThunkAction => {
  return async (dispatch: Dispatch<*>) => {
    if (shouldShowLoader) {
      dispatch(showLoader());
    }

    logger.log('action: fetchPublicationStart');

    const publication = await getPublicationDetail({ id });
    const card = transformPublicationToCard(publication);
    logger.log('action: fetchPublication');

    await Promise.all([
      dispatch(updateCard(card)),
      dispatch(updatePublication(publication)),
    ]);

    logger.log('action: fetchPublicationEnd');
    dispatch(hideLoader());
  };
};

export const toggleList = (request: ToggleListRequest): ThunkAction => {
  return async (dispatch: Dispatch<*>, getState: Function) => {
    logger.log('action: toggleListStart');
    dispatch(showLoader());

    const publication = await postToggleList(request);
    const card = transformPublicationToCard(publication);
    const { cards } = getState().toJS().card;
    logger.log('action: toggleList');

    if (cards) {
      const newCards = cards ? cloneObjectArray(cards) : [];
      const index = findIndexById(newCards, card.id);
      newCards[index] = card;
      await dispatch(updateCards(newCards));
    }

    await Promise.all([
      dispatch(updateCard(card)),
      dispatch(updatePublication(publication)),
    ]);

    logger.log('action: toggleListEnd');
    dispatch(hideLoader());
  };
};
