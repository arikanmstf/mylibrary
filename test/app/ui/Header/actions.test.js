import { mapStateToProps, mapDispatchToProps } from 'ui/Header/actions';
import * as screenActions from 'ui/Screen/actions';
import { createFakeStore, dispatch } from 'helpers/connect/fakeStore';

const state = createFakeStore({ screen: { isDrawerOpen: 'mockedValue' } }).getState();

describe('test/app/ui/Header/actions.test.js', () => {
  it('mapDispatchToProps', () => {
    screenActions.showDrawer = jest.fn();
    screenActions.showDrawer.mockClear();
    expect(mapDispatchToProps(dispatch)).toMatchSnapshot();
  });

  it('mapStateToProps', () => {
    expect(mapStateToProps(state)).toMatchSnapshot();
  });
});
