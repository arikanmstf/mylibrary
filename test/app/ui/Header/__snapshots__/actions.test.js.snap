// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/app/ui/Header/actions.test.js mapDispatchToProps 1`] = `
Object {
  "hideDrawer": [Function],
  "showDrawer": [Function],
}
`;

exports[`test/app/ui/Header/actions.test.js mapStateToProps 1`] = `
Object {
  "isDrawerOpen": "mockedValue",
}
`;
